applications/system/js_app/modules/js_i2c.c

```diff
@@ -0,0 +1,254 @@
+        ret_bad_args(mjs, "Addr must be a number");
+        return;
+    }
+    uint32_t addr = mjs_get_int32(mjs, addr_arg);
+
+    mjs_val_t data_arg = mjs_arg(mjs, 1);
+    if(!mjs_is_array(data_arg)) {
+        ret_bad_args(mjs, "Data must be an array");
+        return;
+    }
+    size_t data_len = mjs_array_length(mjs, data_arg);
+    if(data_len == 0) {
+        ret_bad_args(mjs, "Data array must not be empty");
+        return;
+    }
+    uint8_t* data = malloc(data_len);
+    for(size_t i = 0; i < data_len; i++) {
+        mjs_val_t val = mjs_array_get(mjs, data_arg, i);
+        if(!mjs_is_number(val)) {
+            ret_bad_args(mjs, "Data array must contain only numbers");
+            free(data);
+            return;
+        }
+        data[i] = mjs_get_int32(mjs, val);
+    }
+
+    uint32_t timeout = 1;
+    if(mjs_nargs(mjs) > 2) { // Timeout is optional argument
+        mjs_val_t timeout_arg = mjs_arg(mjs, 2);
+        if(!mjs_is_number(timeout_arg)) {
+            ret_bad_args(mjs, "Timeout must be a number");
+            free(data);
+            return;
+        }
+        timeout = mjs_get_int32(mjs, timeout_arg);
+    }
+
+    furi_hal_i2c_acquire(&furi_hal_i2c_handle_external);
+    bool result = furi_hal_i2c_tx(&furi_hal_i2c_handle_external, addr, data, data_len, timeout);
+    furi_hal_i2c_release(&furi_hal_i2c_handle_external);
+
+    free(data);
+    mjs_return(mjs, mjs_mk_boolean(mjs, result));
+}
+
+static void js_i2c_read(struct mjs* mjs) {
+    if(!check_arg_count_range(mjs, 2, 3)) return;
+
+    mjs_val_t addr_arg = mjs_arg(mjs, 0);
+    if(!mjs_is_number(addr_arg)) {
+        ret_bad_args(mjs, "Addr must be a number");
+        return;
+    }
+    uint32_t addr = mjs_get_int32(mjs, addr_arg);
+
+    mjs_val_t length_arg = mjs_arg(mjs, 1);
+    if(!mjs_is_number(length_arg)) {
+        ret_bad_args(mjs, "Length must be a number");
+        return;
+    }
+    size_t len = mjs_get_int32(mjs, length_arg);
+    if(len == 0) {
+        ret_bad_args(mjs, "Length must not zero");
+        return;
+    }
+
+    uint8_t* mem_addr = malloc(len);
+    memset(mem_addr, 0, len);
+
+    uint32_t timeout = 1;
+    if(mjs_nargs(mjs) > 2) { // Timeout is optional argument
+        mjs_val_t timeout_arg = mjs_arg(mjs, 2);
+        if(!mjs_is_number(timeout_arg)) {
+            ret_bad_args(mjs, "Timeout must be a number");
+            free(mem_addr);
+            return;
+        }
+        timeout = mjs_get_int32(mjs, timeout_arg);
+    }
+
+    furi_hal_i2c_acquire(&furi_hal_i2c_handle_external);
+    uint8_t* read_buffer = malloc(3);
+    bool result = furi_hal_i2c_rx(&furi_hal_i2c_handle_external, addr, mem_addr, len, timeout);
+    furi_hal_i2c_release(&furi_hal_i2c_handle_external);
+
+    mjs_val_t ret = mjs_mk_array(mjs);
+    if(result) {
+        for(size_t i = 0; i < len; i++) {
+            mjs_array_push(mjs, ret, mjs_mk_number(mjs, mem_addr[i]));
+        }
+    }
+    free(mem_addr);
```

Comment:
this one could benefit from just using arraybuf, saves from copyign the buffer twice and gives a more sane return type to user